### ISSUE TITLE
support mkfs option

### ISSUE THREAD

**Issue by emanuc** [calamares/calamares#1424]
[Btrfs from kernel 5.5 supports new checksumm algorithms](https://kdave.github.io/btrfs-hilights-5.5-new-hashes/), to select it must be done when formatting (**mkfs.btrfs --csum xxhash /dev/sda**), it would be convenient to have possibility in the installer to include an option for mkfs.
The openSUSE installer will include it soon.
https://pbs.twimg.com/media/EZgymOaUYAASheN?format=png&name=900x900
It is feasible?
Thanks to anyone who will take the time to answer :-)

**Comment by kkofler** [comment]
This feature was already requested, please see issue #1416. Hence, I am closing this issue as a duplicate of #1416.

**Comment by emanuc** [comment]
I'm not sure it's the same feature request.
#1416 is a request to select mount options after formatting in btrfs.
Here I am putting forward an idea about the formatting options on btrfs because they cannot be changed after formatting.
Maybe if I select Btrfs, a list of mkfs with the checksum options available, in this case:
crc32 (is the default when formatting a partition in Btrfs)
XXHASH
SHA256
BLAKE2B
In the future when authentication will be implemented:
`mkfs.btrfs --csum "hmac (sha256)" --auth-key 0123456 /dev/disk`
[https://patchwork.kernel.org/patch/11548387/](url)

**Comment by kkofler** [comment]
You're right, there's the `mkfs` options and there's the runtime mount options.

**Comment by abucodonosor** [comment]
All those options need UI support, otherwise, they are all meaningless.

In this case or BTRFS in general with all the craziness, they add/remove/half implement
a whole new BTRFS UI module would be needed.

**Comment by abucodonosor** [comment]
And btw before all those options *can* be implemented not only for BTRFS, UI support for choosing the FS is needed.

**Comment by mirh** [comment]
This would be also useful for F2FS and compression. 
AFAIU your require `extra_attr,compression` to use it.

**Comment by adriaandegroot** [comment]
Hacktoberfest 2020 has not yet started, but this is a good kind of issue to do for it.

## Procedural

- [ ] Make sure you can compile Calamares locally (see https://github.com/calamares/calamares/blob/calamares/CONTRIBUTING.md, which will point you onwards to the wiki)
- [ ] Make sure the *partition* module is enabled in your Calamares build (by default it is compiled, and also enabled in settings.conf: check the CMake output to see that it will be compiled, and check the settings.conf in your build directory to make the module show up). Take note that the *partition* module is one that **will** destroy data on your system! Use a VM to be safe.
- [ ] Your system will need a recent KPMCore.
- [ ] Module sources are in https://github.com/calamares/calamares/tree/calamares/src/modules/partition ; make a branch, add your copyright statement to one of the .cpp files and add your name to top-level CHANGES, then start a PR with WIP in the title

## Functional

You can leverage some of the work already done by Gael Portay, who added a *features* field to custom partitioning layouts. This is all invisible to the user; there is no UI and for the time being, leave it without UI (all kinds of power-user features are out-of-scope for Calamares unless we can design a really good not-in-your-face UI for them; see the *Extras* section below).

- [ ] Find in the partitioning module where filesystems are created (presumably a *CreateFileSystemJob*).
- [ ] Find the API for adding flags / features to the create-job (KPMCore probably has an API to attach extra flags; if it doesn't, head on over to [KPMCore on KDE Invent](https://invent.kde.org/system/kpmcore) and work there, first (does not participate in hacktoberfest)
- [ ] Possibly extend the custom partitioning configuration with custom FS flags (I don't know just now if the existing flags apply to the partition, or to the FS created on the partition, or both)
- [ ] Add a new map, say *fsflags*, that map filesystem names to lists of flags / feature values to set on that filesystem. Make sure you use the canonical untranslated names (there is code in Calamares somewhere to handle **some** variations of FS names; see `PartUtils.h`)
- [ ] When creating the filesystem, collect flags from the partition layout and general FS-to-flags maps, and apply them.

## Extras

If might be interesting to offer a UI; something that offers a checklist of known options, and offers the user to know better and add other flags. That's fairly FS-specific, though, so will require trawling through manpages to collect them. KPMCore might have such lists already, though, see the FileSystem header and classes.

A proof-of-concept with an in-your-face UI would be ok, although we'd probably CSS-style it to invisibility for now.

## Post-Procedural

Feel free to push to the PR branch regularly; this shows how your progress is and allows early comments. Please stick to the commit-message style used in the rest of Calamares; do write out what-and-why with commits. It's OK to leave fixme's around for later in the branch and we can coordinate force-pushes if need be.
